File: app/src/main/java/org/transdroid/daemon/Rtorrent/RtorrentAdapter.java
/*
 *	This file is part of Transdroid <http://www.transdroid.org>
 *	
 *	Transdroid is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *	
 *	Transdroid is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *	
 *	You should have received a copy of the GNU General Public License
 *	along with Transdroid.  If not, see <http://www.gnu.org/licenses/>.
 *	
 */
package org.transdroid.daemon.Rtorrent;

import org.transdroid.core.gui.log.Log;
import org.transdroid.daemon.Daemon;
import org.transdroid.daemon.DaemonException;
import org.transdroid.daemon.DaemonException.ExceptionType;
import org.transdroid.daemon.DaemonSettings;
import org.transdroid.daemon.IDaemonAdapter;
import org.transdroid.daemon.Label;
import org.transdroid.daemon.Priority;
import org.transdroid.daemon.Torrent;
import org.transdroid.daemon.TorrentDetails;
import org.transdroid.daemon.TorrentFile;
import org.transdroid.daemon.TorrentStatus;
import org.transdroid.daemon.task.AddByFileTask;
import org.transdroid.daemon.task.AddByMagnetUrlTask;
import org.transdroid.daemon.task.AddByUrlTask;
import org.transdroid.daemon.task.DaemonTask;
import org.transdroid.daemon.task.DaemonTaskFailureResult;
import org.transdroid.daemon.task.DaemonTaskResult;
import org.transdroid.daemon.task.DaemonTaskSuccessResult;
import org.transdroid.daemon.task.GetFileListTask;
import org.transdroid.daemon.task.GetFileListTaskSuccessResult;
import org.transdroid.daemon.task.GetTorrentDetailsTask;
import org.transdroid.daemon.task.GetTorrentDetailsTaskSuccessResult;
import org.transdroid.daemon.task.RemoveTask;
import org.transdroid.daemon.task.RetrieveTask;
import org.transdroid.daemon.task.RetrieveTaskSuccessResult;
import org.transdroid.daemon.task.SetFilePriorityTask;
import org.transdroid.daemon.task.SetLabelTask;
import org.transdroid.daemon.task.SetTransferRatesTask;
import org.transdroid.daemon.util.HttpHelper;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import de.timroes.axmlrpc.XMLRPCClient;
import de.timroes.axmlrpc.XMLRPCClient.UnauthorizdException;
import de.timroes.axmlrpc.XMLRPCException;

/**
 * An adapter that allows for easy access to rTorrent torrent data. Communication is handled via the XML-RPC protocol as
 * implemented by the aXMLRPC library.
 * @author erickok
 */
public class RtorrentAdapter implements IDaemonAdapter {

	private static final String LOG_NAME = "rTorrent daemon";

	private static final String DEFAULT_RPC_URL = "/RPC2";

	private DaemonSettings settings;
	private XMLRPCClient rpcclient;
	private List<Label> lastKnownLabels = null;

	public RtorrentAdapter(DaemonSettings settings) {
		this.settings = settings;
	}

	@Override
	public DaemonTaskResult executeTask(Log log, DaemonTask task) {

		try {
			switch (task.getMethod()) {
				case Retrieve:

					// @formatter:off
				Object result = makeRtorrentCall(log,"d.multicall2",
						new String[] { "", "main",
						"d.hash=",
						"d.name=",
						"d.state=",
						"d.down.rate=",
						"d.up.rate=",
						"d.peers_connected=",
						"d.peers_not_connected=",
						"d.peers_accounted=",
						"d.bytes_done=",
						"d.up.total=",
						"d.size_bytes=",
						"d.creation_date=",
						"d.left_bytes=",
						"d.complete=",
						"d.is_active=", 
						"d.is_hash_checking=", 
						"d.base_path=",
						"d.base_filename=",
						"d.message=",
						"d.custom=addtime",
						"d.custom=seedingtime",
						"d.custom1=",
						"d.peers_complete=",
						"d.peers_accounted=" });
				// @formatter:on
					return new RetrieveTaskSuccessResult((RetrieveTask) task, onTorrentsRetrieved(result),
							lastKnownLabels);

				case GetTorrentDetails:

					// @formatter:off
				Object dresult = makeRtorrentCall(log,"t.multicall", new String[] {
						task.getTargetTorrent().getUniqueID(),
						"", 
						"t.url=" });
				// @formatter:on
					return new GetTorrentDetailsTaskSuccessResult((GetTorrentDetailsTask) task,
							onTorrentDetailsRetrieved(log, dresult));

				case GetFileList:

					// @formatter:off
				Object fresult = makeRtorrentCall(log,"f.multicall", new String[] {
						task.getTargetTorrent().getUniqueID(),
						"", 
						"f.path=",
						"f.size_bytes=",
						"f.priority=",
						"f.completed_chunks=",
						"f.size_chunks=",
						"f.priority=",
						"f.frozen_path=" });
				// @formatter:on
					return new GetFileListTaskSuccessResult((GetFileListTask) task,
							onTorrentFilesRetrieved(fresult, task.getTargetTorrent()));

				case AddByFile:

					// Request to add a torrent by local .torrent file
					File file = new File(URI.create(((AddByFileTask) task).getFile()));
					FileInputStream in = new FileInputStream(file);
					ByteArrayOutputStream baos = new ByteArrayOutputStream();
					byte[] buffer = new byte[(int) file.length()];
					int read;
					while ((read = in.read(buffer, 0, buffer.length)) > 0) {
						baos.write(buffer, 0, read);
					}
					byte[] bytes = baos.toByteArray();
					int size = (int) file.length() * 2;
					final int XMLRPC_EXTRA_PADDING = 1280;
					makeRtorrentCall(log, "network.xmlrpc.size_limit.set", new Object[]{size + XMLRPC_EXTRA_PADDING});
					makeRtorrentCall(log, "load.raw_start", new Object[]{bytes});
					return new DaemonTaskSuccessResult(task);

				case AddByUrl:

					// Request to add a torrent by URL
					String url = ((AddByUrlTask) task).getUrl();
					makeRtorrentCall(log, "load.start", new String[]{"",url});
					return new DaemonTaskSuccessResult(task);

				case AddByMagnetUrl:

					// Request to add a magnet link by URL
					String magnet = ((AddByMagnetUrlTask) task).getUrl();
					makeRtorrentCall(log, "load.start", new String[]{"",magnet});
					return new DaemonTaskSuccessResult(task);

				case Remove:

					// Remove a torrent
					RemoveTask removeTask = (RemoveTask) task;
					if (removeTask.includingData()) {
						makeRtorrentCall(log, "d.custom5.set",
								new String[]{task.getTargetTorrent().getUniqueID(), "1"});
					}
					makeRtorrentCall(log, "d.erase", new String[]{task.getTargetTorrent().getUniqueID()});
					return new DaemonTaskSuccessResult(task);

				case Pause:

					// Pause a torrent
					makeRtorrentCall(log, "d.pause", new String[]{task.getTargetTorrent().getUniqueID()});
					return new DaemonTaskSuccessResult(task);

				case PauseAll:

					// Resume all torrents
					makeRtorrentCall(log, "d.multicall2", new String[]{"","main", "d.pause="});
					return new DaemonTaskSuccessResult(task);

				case Resume:

					// Resume a torrent
					makeRtorrentCall(log, "d.resume", new String[]{task.getTargetTorrent().getUniqueID()});
					return new DaemonTaskSuccessResult(task);

				case ResumeAll:

					// Resume all torrents
					makeRtorrentCall(log, "d.multicall2", new String[]{"", "main", "d.resume="});
					return new DaemonTaskSuccessResult(task);

				case Stop:

					// Stop a torrent
					makeRtorrentCall(log, "d.stop", new String[]{task.getTargetTorrent().getUniqueID()});
					return new DaemonTaskSuccessResult(task);

				case StopAll:

					// Stop all torrents
					makeRtorrentCall(log, "d.multicall2", new String[]{"", "main", "d.stop="});
					return new DaemonTaskSuccessResult(task);

				case Start:

					// Start a torrent
					makeRtorrentCall(log, "d.start", new String[]{task.getTargetTorrent().getUniqueID()});
					return new DaemonTaskSuccessResult(task);

				case StartAll:

					// Start all torrents
					makeRtorrentCall(log, "d.multicall2", new String[]{"", "main", "d.start="});
					return new DaemonTaskSuccessResult(task);

				case SetFilePriorities:

					// For each of the chosen files belonging to some torrent, set the priority
					SetFilePriorityTask prioTask = (SetFilePriorityTask) task;
					String newPriority = "" + convertPriority(prioTask.getNewPriority());
					// One at a time; rTorrent doesn't seem to support a multicall on a selective number of files
					for (TorrentFile forFile : prioTask.getForFiles()) {
						makeRtorrentCall(log, "f.priority.set",
								new String[]{task.getTargetTorrent().getUniqueID() + ":f" + forFile.getKey(),
										newPriority});
					}
					return new DaemonTaskSuccessResult(task);

				case SetTransferRates:

					// Request to set the maximum transfer rates
					SetTransferRatesTask ratesTask = (SetTransferRatesTask) task;
					makeRtorrentCall(log, "throttle.global_down.max_rate.set", new String[]{"",(ratesTask.getDownloadRate() == null ? "0" :
							ratesTask.getDownloadRate().toString() + "k")});
					makeRtorrentCall(log, "throttle.global_up.max_rate.set", new String[]{"",
							(ratesTask.getUploadRate() == null ? "0" : ratesTask.getUploadRate().toString() + "k")});
					return new DaemonTaskSuccessResult(task);

				case SetLabel:

					SetLabelTask labelTask = (SetLabelTask) task;
					makeRtorrentCall(log, "d.custom1.set",
							new String[]{task.getTargetTorrent().getUniqueID(), labelTask.getNewLabel()});
					return new DaemonTaskSuccessResult(task);

				case ForceRecheck:

					// Force re-check of data of a torrent
					makeRtorrentCall(log, "d.check_hash", new String[]{task.getTargetTorrent().getUniqueID()});
					return new DaemonTaskSuccessResult(task);

				default:
					return new DaemonTaskFailureResult(task, new DaemonException(ExceptionType.MethodUnsupported,
							task.getMethod() + " is not supported by " + getType()));
			}
		} catch (DaemonException e) {
			return new DaemonTaskFailureResult(task, e);
		} catch (FileNotFoundException e) {
			return new DaemonTaskFailureResult(task, new DaemonException(ExceptionType.FileAccessError, e.toString()));
		} catch (IOException e) {
			return new DaemonTaskFailureResult(task, new DaemonException(ExceptionType.ConnectionError, e.toString()));
		}
	}

	private Object makeRtorrentCall(Log log, String serverMethod, Object[] arguments)
			throws DaemonException, MalformedURLException {

		// Initialise the HTTP client
		if (rpcclient == null) {
			initialise();
		}

		String params = "";
		for (Object arg : arguments) {
			params += " " + arg.toString();
		}
		try {
			log.d(LOG_NAME, "Calling " + serverMethod + " with params [" +
							(params.length() > 100 ? params.substring(0, 100) + "..." : params) + " ]");
			return rpcclient.call(serverMethod, arguments);
		} catch (XMLRPCException e) {
			log.d(LOG_NAME, e.toString());
			if (e.getCause() instanceof UnauthorizdException) {
				throw new DaemonException(ExceptionType.AuthenticationFailure, e.toString());
			}
			if (e.getCause() instanceof DaemonException) {
				throw (DaemonException) e.getCause();
			}
			throw new DaemonException(ExceptionType.ConnectionError,
					"Error making call to " + serverMethod + " with params [" +
							(params.length() > 100 ? params.substring(0, 100) + "..." : params) + " ]: " +
							e.toString());
		}

	}

	/**
	 * Instantiates a XML-RPC client with proper credentials.
	 * @throws DaemonException On conflicting settings (i.e. user authentication but no password or username provided)
	 * @throws MalformedURLException Thrown when the URL could not be properly constructed
	 */
	private void initialise() throws DaemonException, MalformedURLException {

		int flags = XMLRPCClient.FLAGS_8BYTE_INT;
		this.rpcclient = new XMLRPCClient(HttpHelper.createStandardHttpClient(settings, true), buildWebUIUrl(), flags);

	}

	/**
	 * Build the URL of rTorrent's XML-RPC location from the user settings.
	 * @return The URL of the RPC API
	 */
	private String buildWebUIUrl() {
		return (settings.getSsl() ? "https://" : "http://") + settings.getAddress() + ":" + settings.getPort() +
				(settings.getFolder() == null || settings.getFolder().equals("") ? DEFAULT_RPC_URL :
						settings.getFolder());
	}

	private List<Torrent> onTorrentsRetrieved(Object response) throws DaemonException {

		if (response == null || !(response instanceof Object[])) {

			throw new DaemonException(ExceptionType.ParsingFailed,
					"Response on retrieveing torrents did not return a list of objects");

		} else {

			// Parse torrent list from response
			// Formatted as Object[][], see http://libtorrent.rakshasa.no/wiki/RTorrentCommands#Download
			List<Torrent> torrents = new ArrayList<Torrent>();
			Map<String, Integer> labels = new HashMap<String, Integer>();
			Object[] responseList = (Object[]) response;
			for (int i = 0; i < responseList.length; i++) {

				Object[] info = (Object[]) responseList[i];
				String error = (String) info[18];
				error = error.equals("") ? null : error;

				// Determine the time added
				Date added;
				Long addtime = null;
				try {
					addtime = Long.valueOf(((String) info[19]).trim());
				} catch (NumberFormatException e) {
					// Not a number (timestamp); ignore and fall back to using creationtime
				}
				if (addtime != null)
				// Successfully received the addtime from rTorrent (which is a String like '1337089336\n')
				{
					added = new Date(addtime * 1000L);
				} else {
					// rTorrent didn't have the addtime (missing plugin?): base it on creationtime instead
					if (info[11] instanceof Long) {
						added = new Date((Long) info[11] * 1000L);
					} else {
						added = new Date((Integer) info[11] * 1000L);
					}
				}

				// Determine the seeding time
				Date finished = null;
				Long seedingtime = null;
				try {
					seedingtime = Long.valueOf(((String) info[20]).trim());
				} catch (NumberFormatException e) {
					// Not a number (timestamp); ignore and fall back to using creationtime
				}
				if (seedingtime != null)
				// Successfully received the seedingtime from rTorrent (which is a String like '1337089336\n')
				{
					finished = new Date(seedingtime * 1000L);
				}

				// Determine the label
				String label = null;
				try {
					label = URLDecoder.decode((String) info[21], "UTF-8");
					if (labels.containsKey(label)) {
						labels.put(label, labels.get(label) + 1);
					} else {
						labels.put(label, 0);
					}
				} catch (UnsupportedEncodingException e) {
					// Can't decode label name; ignore it
				}

				if (info[3] instanceof Long) {

					// rTorrent uses the i8 dialect which returns 64-bit integers
					long rateDownload = (Long) info[3];
					String basePath = (String) info[16];

					// @formatter:off
					torrents.add(new Torrent(
						i,
						(String)info[0], // hash
						(String)info[1], // name
						convertTorrentStatus((Long)info[2], (Long)info[13], (Long)info[14], (Long)info[15]), // status
						basePath.substring(0, basePath.indexOf((String)info[17])), // locationDir
						((Long)info[3]).intValue(), // rateDownload
						((Long)info[4]).intValue(), // rateUpload
						((Long)info[22]).intValue(), // seedersConnected
						((Long)info[5]).intValue() + ((Long)info[6]).intValue(), // seedersKnown
						((Long)info[23]).intValue(), // leechersConnected
						((Long)info[5]).intValue() + ((Long)info[6]).intValue(), // leechersKnown
						(rateDownload > 0? (int) (((Long)info[12]) / rateDownload): -1), // eta (bytes left / rate download, if rate > 0)
						(Long)info[8], // downloadedEver
						(Long)info[9], // uploadedEver
						(Long)info[10], // totalSize
						((Long)info[8]).floatValue() / ((Long)info[10]).floatValue(), // partDone
						0f, // TODO: Add availability data
						label,
						added,
						finished,
						error,
						settings.getType()));
					// @formatter:on

				} else {

					// rTorrent uses the default dialect with 32-bit integers
					int rateDownload = (Integer) info[3];
					String basePath = (String) info[16];

					// @formatter:off
					torrents.add(new Torrent(
						i,
						(String)info[0], // hash
						(String)info[1], // name
						convertTorrentStatus(((Integer)info[2]).longValue(), ((Integer)info[13]).longValue(), ((Integer)info[14]).longValue(), ((Integer)info[15]).longValue()), // status
						basePath.substring(0, basePath.indexOf((String)info[17])), // locationDir
						rateDownload, // rateDownload
						(Integer)info[4], // rateUpload
						(Integer)info[22], // seedersConnected
						(Integer)info[5] + (Integer)info[6], // seedersKnown
						(Integer)info[23], // leechersConnected
						(Integer)info[5] + (Integer)info[6], // leechersKnown
						(rateDownload > 0? (Integer)info[12] / rateDownload: -1), // eta (bytes left / rate download, if rate > 0)
						(Integer)info[8], // downloadedEver
						(Integer)info[9], // uploadedEver
						(Integer)info[10], // totalSize
						((Integer)info[8]).floatValue() / ((Integer)info[10]).floatValue(), // partDone
						0f, // TODO: Add availability data
						label,
						added,
						finished,
						error,
						settings.getType()));
					// @formatter:on

				}
			}
			lastKnownLabels = new ArrayList<Label>();
			for (Entry<String, Integer> pair : labels.entrySet()) {
				if (pair.getKey() != null) {
					lastKnownLabels.add(new Label(pair.getKey(), pair.getValue()));
				}
			}
			return torrents;

		}

	}

	private List<TorrentFile> onTorrentFilesRetrieved(Object response, Torrent torrent) throws DaemonException {

		if (response == null || !(response instanceof Object[])) {

			throw new DaemonException(ExceptionType.ParsingFailed,
					"Response on retrieveing torrent files did not return a list of objects");

		} else {

			// Parse torrent files from response
			// Formatted as Object[][], see http://libtorrent.rakshasa.no/wiki/RTorrentCommands#Download
			List<TorrentFile> files = new ArrayList<TorrentFile>();
			Object[] responseList = (Object[]) response;
			for (int i = 0; i < responseList.length; i++) {

				Object[] info = (Object[]) responseList[i];
				if (info[1] instanceof Long) {

					// rTorrent uses the i8 dialect which returns 64-bit integers
					Long size = (Long) info[1];
					Long chunksDone = (Long) info[3];
					Long chunksTotal = (Long) info[4];
					Long priority = (Long) info[5];

					// @formatter:off
					files.add(new TorrentFile(
						"" + i,
						(String)info[0], // name
						((String)info[6]).substring(torrent.getLocationDir().length()), // relativePath (= fullPath - torrent locationDir)
						(String)info[6], // fullPath
						size, // size
						(long) (size * ((float)chunksDone / (float)chunksTotal)), // done
						convertRtorrentPriority(priority.intValue()))); // priority
					// @formatter:on

				} else {

					// rTorrent uses the default dialect with 32-bit integers
					Integer size = (Integer) info[1];
					Integer chunksDone = (Integer) info[3];
					Integer chunksTotal = (Integer) info[4];
					Integer priority = (Integer) info[5];

					// @formatter:off
					files.add(new TorrentFile(
						"" + i,
						(String)info[0], // name
						((String)info[6]).substring(torrent.getLocationDir().length()), // relativePath (= fullPath - torrent locationDir)
						(String)info[0], // fullPath
						size, // size
						(int) (size * ((float)chunksDone / (float)chunksTotal)), // done
					convertRtorrentPriority(priority))); // priority
					// @formatter:on

				}
			}
			return files;

		}

	}

	private Priority convertRtorrentPriority(int code) {
		// Note that Rtorrent has no low priority value
		switch (code) {
			case 0:
				return Priority.Off;
			case 2:
				return Priority.High;
			default:
				return Priority.Normal;
		}
	}

	private int convertPriority(Priority priority) {
		// Note that Rtorrent has no low priority value
		switch (priority) {
			case Off:
				return 0;
			case High:
				return 2;
			default:
				return 1;
		}
	}

	private TorrentStatus convertTorrentStatus(Long state, Long complete, Long active, Long checking) {
		if (state == 0) {
			return TorrentStatus.Queued;
		} else if (active == 1) {
			if (complete == 1) {
				return TorrentStatus.Seeding;
			} else {
				return TorrentStatus.Downloading;
			}
		} else if (checking == 1) {
			return TorrentStatus.Checking;
		} else {
			return TorrentStatus.Paused;
		}
	}

	private TorrentDetails onTorrentDetailsRetrieved(Log log, Object response) throws DaemonException {

		if (response == null || !(response instanceof Object[])) {

			throw new DaemonException(ExceptionType.ParsingFailed,
					"Response on retrieveing trackers did not return a list of objects");

		} else {

			// Parse a torrent's trackers from response
			// Formatted as Object[][], see http://libtorrent.rakshasa.no/wiki/RTorrentCommands#Download
			List<String> trackers = new ArrayList<String>();
			Object[] responseList = (Object[]) response;
			try {
				for (Object aResponseList : responseList) {
					Object[] info = (Object[]) aResponseList;
					trackers.add((String) info[0]);
				}
			} catch (Exception e) {
				log.e(LOG_NAME, e.toString());
			}
			return new TorrentDetails(trackers, null);

		}

	}

	@Override
	public Daemon getType() {
		return settings.getType();
	}

	@Override
	public DaemonSettings getSettings() {
		return this.settings;
	}

}

File: protocol/src/main/kotlin/org/transdroid/protocol/rtorrent/RtorrentAdapter.kt
/*
 * Copyright 2010-2026 Eric Kok et al.
 *
 * Transdroid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Transdroid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Transdroid. If not, see <https://www.gnu.org/licenses/>.
 */
package org.transdroid.protocol.rtorrent

import java.net.URLDecoder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Credentials
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.transdroid.protocol.DaemonAdapter
import org.transdroid.protocol.DaemonConfig
import org.transdroid.protocol.DaemonException
import org.transdroid.protocol.FilePriority
import org.transdroid.protocol.Torrent
import org.transdroid.protocol.TorrentFile
import org.transdroid.protocol.TorrentStatus
import org.transdroid.protocol.Tracker
import org.transdroid.protocol.TrackerStatus
import org.transdroid.protocol.internal.executeOnIo

/**
 * Adapter for rTorrent 0.9.7+ via XML-RPC over HTTP, the common seedbox setup where a web
 * server (or ruTorrent) exposes the SCGI socket at an endpoint like /RPC2. Uses the modern
 * dotted command names (d.multicall2, load.start); the pre-0.9.7 underscore API is not
 * supported.
 */
class RtorrentAdapter(
    override val config: DaemonConfig,
    private val httpClient: OkHttpClient,
) : DaemonAdapter {

    private val rpcUrl = config.baseUrl +
        (config.path?.takeIf { it.isNotBlank() } ?: "/RPC2").let { if (it.startsWith("/")) it else "/$it" }

    override suspend fun testConnection(): String {
        val version = call("system.client_version") as? String ?: "unknown"
        return "rTorrent $version"
    }

    override suspend fun listTorrents(): List<Torrent> {
        val rows = call(
            "d.multicall2",
            "", "main",
            "d.hash=", "d.name=", "d.state=", "d.complete=", "d.is_active=", "d.hashing=",
            "d.down.rate=", "d.up.rate=", "d.size_bytes=", "d.completed_bytes=", "d.up.total=",
            "d.ratio=", "d.peers_connected=", "d.timestamp.started=", "d.directory=", "d.message=",
            "d.custom1=", "d.peers_complete=",
        ) as? List<*> ?: throw DaemonException.UnexpectedResponse("Unexpected d.multicall2 reply")
        return rows.map { row ->
            val fields = row as? List<*> ?: throw DaemonException.UnexpectedResponse("Bad multicall row")
            parseTorrent(fields)
        }
    }

    private fun parseTorrent(fields: List<*>): Torrent {
        fun str(index: Int) = fields.getOrNull(index)?.toString().orEmpty()
        fun num(index: Int) = (fields.getOrNull(index) as? Long) ?: 0L

        val state = num(2)
        val complete = num(3) == 1L
        val isActive = num(4) == 1L
        val hashing = num(5) > 0L
        val downRate = num(6)
        val sizeBytes = num(8)
        val completedBytes = num(9)
        val message = str(15).takeIf { it.isNotBlank() }

        val status = when {
            message != null -> TorrentStatus.ERROR
            hashing -> TorrentStatus.CHECKING
            state == 0L || !isActive -> TorrentStatus.PAUSED
            complete -> TorrentStatus.SEEDING
            else -> TorrentStatus.DOWNLOADING
        }
        val eta = if (status == TorrentStatus.DOWNLOADING && downRate > 0) {
            (sizeBytes - completedBytes) / downRate
        } else {
            null
        }
        return Torrent(
            id = str(0),
            name = str(1),
            status = status,
            progress = if (sizeBytes <= 0) 0f else (completedBytes.toFloat() / sizeBytes).coerceIn(0f, 1f),
            downloadRate = downRate,
            uploadRate = num(7),
            etaSeconds = eta,
            sizeBytes = sizeBytes,
            downloadedBytes = completedBytes,
            uploadedBytes = num(10),
            // d.ratio is reported in per-mille
            ratio = num(11) / 1000f,
            peersConnected = num(12).toInt(),
            // d.peers_complete is the number of connected peers that are seeding (complete)
            seedersConnected = num(17).toInt(),
            leechersConnected = (num(12) - num(17)).toInt().coerceAtLeast(0),
            addedTimestamp = num(13).takeIf { it > 0 },
            downloadDir = str(14).takeIf { it.isNotBlank() },
            error = message,
            // ruTorrent stores its label encodeURIComponent-encoded in custom1; that
            // encoding leaves "+" literal, so protect it from URLDecoder's plus-to-space
            labels = str(16).takeIf { it.isNotBlank() }?.let { raw ->
                listOf(
                    try {
                        URLDecoder.decode(raw.replace("+", "%2B"), "UTF-8")
                    } catch (e: IllegalArgumentException) {
                        raw
                    }
                )
            } ?: emptyList(),
        )
    }

    override suspend fun addByUrl(url: String, startPaused: Boolean) {
        // load.normal loads without starting; load.start loads and starts
        call(if (startPaused) "load.normal" else "load.start", "", url)
    }

    override suspend fun addByFile(fileName: String, contents: ByteArray, startPaused: Boolean) {
        call(if (startPaused) "load.raw" else "load.raw_start", "", contents)
    }

    override suspend fun start(torrentId: String) {
        call("d.start", torrentId)
    }

    override suspend fun pause(torrentId: String) {
        call("d.stop", torrentId)
    }

    override suspend fun remove(torrentId: String, deleteData: Boolean) {
        if (deleteData) {
            // ruTorrent convention: an event hook on custom5 erases the data on removal.
            // Harmless when no such hook is configured; rTorrent itself never deletes data.
            call("d.custom5.set", torrentId, "1")
        }
        call("d.erase", torrentId)
    }

    override suspend fun listFiles(torrentId: String): List<TorrentFile> {
        val rows = call(
            "f.multicall",
            torrentId, "",
            "f.path=", "f.size_bytes=", "f.completed_chunks=", "f.size_chunks=", "f.priority=",
        ) as? List<*> ?: throw DaemonException.UnexpectedResponse("Unexpected f.multicall reply")
        return rows.mapIndexed { index, row ->
            val fields = row as? List<*> ?: throw DaemonException.UnexpectedResponse("Bad multicall row")
            val size = (fields.getOrNull(1) as? Long) ?: 0L
            val completedChunks = (fields.getOrNull(2) as? Long) ?: 0L
            val sizeChunks = (fields.getOrNull(3) as? Long) ?: 0L
            TorrentFile(
                index = index,
                path = fields.getOrNull(0)?.toString().orEmpty(),
                sizeBytes = size,
                downloadedBytes = if (sizeChunks <= 0) 0L else size * completedChunks / sizeChunks,
                priority = when ((fields.getOrNull(4) as? Long) ?: 1L) {
                    0L -> FilePriority.OFF
                    2L -> FilePriority.HIGH
                    else -> FilePriority.NORMAL
                },
            )
        }
    }

    override suspend fun setFilePriority(torrentId: String, fileIndex: Int, priority: FilePriority) {
        // rTorrent priorities: 0 = off, 1 = normal (LOW folds into it), 2 = high
        val value = when (priority) {
            FilePriority.OFF -> 0L
            FilePriority.HIGH -> 2L
            else -> 1L
        }
        call("f.priority.set", "$torrentId:f$fileIndex", value)
        call("d.update_priorities", torrentId)
    }

    override suspend fun listTrackers(torrentId: String): List<Tracker> {
        val rows = call(
            "t.multicall",
            torrentId, "",
            "t.url=", "t.is_enabled=", "t.success_counter=", "t.failed_counter=",
            "t.scrape_complete=", "t.scrape_incomplete=",
        ) as? List<*> ?: throw DaemonException.UnexpectedResponse("Unexpected t.multicall reply")
        return rows.map { row ->
            val fields = row as? List<*> ?: throw DaemonException.UnexpectedResponse("Bad multicall row")
            fun str(index: Int) = fields.getOrNull(index)?.toString().orEmpty()
            fun num(index: Int) = (fields.getOrNull(index) as? Long) ?: 0L
            val enabled = num(1) != 0L
            val status = when {
                !enabled -> TrackerStatus.IDLE
                num(2) > 0L -> TrackerStatus.WORKING
                num(3) > 0L -> TrackerStatus.ERROR
                else -> TrackerStatus.IDLE
            }
            Tracker(
                url = str(0),
                status = status,
                seeders = num(4).toInt().takeIf { it >= 0 },
                leechers = num(5).toInt().takeIf { it >= 0 },
            )
        }
    }

    private suspend fun call(method: String, vararg params: Any?): Any? {
        val builder = Request.Builder()
            .url(rpcUrl)
            .post(XmlRpc.buildRequest(method, params.toList()).toRequestBody("text/xml".toMediaType()))
        val username = config.username
        if (!username.isNullOrEmpty()) {
            builder.header("Authorization", Credentials.basic(username, config.password.orEmpty()))
        }
        httpClient.executeOnIo(builder.build()).use { response ->
            when {
                response.code == 401 || response.code == 403 ->
                    throw DaemonException.Authentication("rTorrent's web server rejected the username/password")
                !response.isSuccessful ->
                    throw DaemonException.UnexpectedResponse("rTorrent returned HTTP ${response.code}")
            }
            val body = response.body ?: throw DaemonException.UnexpectedResponse("Empty rTorrent response")
            // executeOnIo only wraps the round-trip up to receiving headers; the body still
            // streams off the same connection, so parsing it can still block on the socket
            // and must stay on IO too, or it risks a NetworkOnMainThreadException on whatever
            // dispatcher called the adapter.
            return withContext(Dispatchers.IO) { XmlRpc.parseResponse(body.byteStream()) }
        }
    }
}
